json2junit/json2junit.go

```diff
@@ -0,0 +1,315 @@
+	TestCases []*junitTestCase
+}
+
+func (s junitTestSuite) testCase(name string) *junitTestCase {
+	for _, tc := range s.TestCases {
+		if tc.Name == name {
+			return tc
+		}
+	}
+	return nil
+}
+
+type systemOut struct {
+	XMLName xml.Name `xml:"system-out"`
+	Content []byte   `xml:",cdata"`
+}
+
+type junitTestCase struct {
+	XMLName   xml.Name     `xml:"testcase"`
+	Classname string       `xml:"classname,attr"`
+	Name      string       `xml:"name,attr"`
+	Time      float64      `xml:"time,attr"`
+	Result    *junitResult `xml:",omitempty"`
+	systemOut []byte       // Placeholder for the event output.
+}
+
+type junitResult struct {
+	XMLName xml.Name
+	Message string `xml:"message,attr"`
+	Content []byte `xml:",cdata"`
+}
+
+// suite returns the suite with the given name, or nil if there is none.
+func (c *Converter) suite(name string) (int, *junitTestSuite) {
+	for i, s := range c.suites {
+		if s.Name == name {
+			return i, s
+		}
+	}
+	return 0, nil
+}
+
+// processLine converts a single JSON test line.
+func (c *Converter) processJSONEntry(entry jsonEntry) error {
+	switch entry.Action {
+	case "start":
+		// start a new test suite.
+		if _, suite := c.suite(entry.Package); suite != nil {
+			return fmt.Errorf("duplicate start entry for %v", entry.Package)
+		}
+		suite := &junitTestSuite{
+			Name:      entry.Package,
+			Timestamp: entry.Time.Format(time.RFC3339),
+		}
+		c.suites = append(c.suites, suite)
+	case "run":
+		// start a new test case.
+		_, suite := c.suite(entry.Package)
+		if suite == nil {
+			return fmt.Errorf("no start entry for %v", entry.Package)
+		}
+		testCase := suite.testCase(entry.Test)
+		if testCase == nil {
+			suite.Tests++
+			testCase = &junitTestCase{
+				Name:      entry.Test,
+				Classname: entry.Package,
+			}
+			suite.TestCases = append(suite.TestCases, testCase)
+		}
+	case "output":
+		// append output to the current test case or suite.
+		_, suite := c.suite(entry.Package)
+		if suite == nil {
+			return fmt.Errorf("no start entry for %v", entry.Package)
+		}
+		out, ok := strings.CutPrefix(entry.Output, "    ")
+		if !ok {
+			// Ignore all lines that don't start with "    " (4 spaces),
+			// because they are not part of the test output but to the
+			// Go test runner.
+			break
+		}
+		if entry.Test == "" {
+			if suite.SystemOut == nil {
+				suite.SystemOut = new(systemOut)
+			}
+			suite.SystemOut.Content = append(suite.SystemOut.Content, out...)
+		} else {
+			testCase := suite.testCase(entry.Test)
+			if testCase == nil {
+				return fmt.Errorf("no run entry for %v", entry.Test)
+			}
+			testCase.systemOut = append(testCase.systemOut, out...)
+		}
+	case "pass", "skip", "fail":
+		// finish the current test case or suite.
+		i, suite := c.suite(entry.Package)
+		if suite == nil {
+			return fmt.Errorf("no start entry for %v", entry.Package)
+		}
+		if entry.Test == "" {
+			suite.Time = entry.Elapsed
+			if entry.Action != "fail" {
+				suite.SystemOut = nil
```

Comment:
This is the first time a SystemOut is cleared out, so I think it deserves a comment too.

But that raises a question... should this really be cleared out upon skip, too? (Or can a suite not be skipped? Not sure if I have the terminology lined up in my head.)

```suggestion
				// In case of success, we don't care about the output.
				suite.SystemOut = nil
```